plugin/trino-paimon/src/main/java/io/trino/plugin/paimon/PaimonMergeSink.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.paimon;
+
+import io.airlift.slice.Slice;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.connector.ConnectorMergeSink;
+import io.trino.spi.connector.ConnectorPageSink;
+import io.trino.spi.type.TinyintType;
+import org.apache.paimon.types.RowKind;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.IntStream;
+
+/**
+ * Trino {@link ConnectorMergeSink}.
+ */
+public class PaimonMergeSink
+        implements ConnectorMergeSink
+{
+    private final PaimonPageSink pageSink;
+    private final int dataColumnCount;
+
+    public PaimonMergeSink(ConnectorPageSink pageSink, int dataColumnCount)
+    {
+        this.pageSink = (PaimonPageSink) pageSink;
+        this.dataColumnCount = dataColumnCount;
+    }
+
+    @Override
+    public void storeMergedRows(Page page)
+    {
+        int inputChannelCount = page.getChannelCount();
+        if (inputChannelCount != dataColumnCount + 2) {
+            throw new IllegalArgumentException(
+                    String.format(
```

Comment:
Use Guava `checkArgument` method instead. 

Note that `String.format` method should be imported statically. I would recommend reading https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md